metrics/release_test.go

```diff
@@ -0,0 +1,221 @@
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/redhat-appstudio/release-service/kcp"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var _ = Describe("Metrics Release", Ordered, func() {
+
+	BeforeAll(func() {
+		// We need to unregister in advance otherwise it breaks with `AlreadyRegisteredError`
+		metrics.Registry.Unregister(ReleaseAttemptRunningSeconds)
+		metrics.Registry.Unregister(ReleaseAttemptConcurrentTotal)
+		metrics.Registry.Unregister(ReleaseAttemptDurationSeconds)
+	})
+
+	var (
+		AttemptRunningSecondsHeader = inputHeader{
+			Name: "release_attempt_running_seconds",
+			Help: "Release durations from the moment the release resource was created til the release is marked as running",
+		}
+		AttemptDurationSecondsHeader = inputHeader{
+			Name: "release_attempt_duration_seconds",
+			Help: "Release durations from the moment the release PipelineRun was created til the release is marked as finished",
+		}
+		AttemptTotalHeader = inputHeader{
+			Name: "release_attempt_total",
+			Help: "Total number of releases processed by the operator",
+		}
+	)
+
+	const (
+		validReleaseReason   = "valid_release_reason"
+		invalidReleaseReason = "invalid_release_reason"
+		strategy             = "nostrategy"
+	)
+
+	var defaultNamespace = kcp.NamespaceReference{
+		Namespace: "default",
+		Workspace: "default",
+	}
+
+	Context("When RegisterNewRelease is called", func() {
+
+		// As we need to share metrics within the Context, we need to use "per Context" `Before`|`After`All()
+		BeforeAll(func() {
+			// Overriding the metric definitions to simplify the tests
+			//
+			// `Help` can't be overriden due to `https://github.com/prometheus/client_golang/blob/83d56b1144a0c2eb10d399e7abbae3333bebc463/prometheus/registry.go#L314`
+			ReleaseAttemptRunningSeconds = prometheus.NewHistogram(
+				prometheus.HistogramOpts{
+					Name:    "release_attempt_running_seconds",
+					Help:    "Release durations from the moment the release resource was created til the release is marked as running",
+					Buckets: []float64{1, 5, 10, 30},
+				},
+			)
+			ReleaseAttemptConcurrentTotal = prometheus.NewGauge(
+				prometheus.GaugeOpts{
+					Name: "release_attempt_concurrent_requests",
+					Help: "Total number of concurrent release attempts",
+				},
+			)
+			metrics.Registry.MustRegister(ReleaseAttemptRunningSeconds,
+				ReleaseAttemptConcurrentTotal)
+
+		})
+
+		AfterAll(func() {
+			metrics.Registry.Unregister(ReleaseAttemptRunningSeconds)
+			metrics.Registry.Unregister(ReleaseAttemptConcurrentTotal)
+
+		})
+
+		// Buckets for operation duration less or equal to 1, 5, 10 and 30 seconds
+		buckets := []float64{1, 5, 10, 30}
```

Comment:
I would declare every single variable in each tests to ensure they don't get modified and their values carried over.